src/kernels/MIOpenUnfold.cpp

```diff
@@ -0,0 +1,238 @@
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename DTYPE>
+__device__ void unfoldForward4D(const DTYPE* __restrict__ input,
+                                DTYPE* __restrict__ output,
+                                int N,
+                                int C,
+                                int H,
+                                int W,
+                                int P,
+                                int L,
+                                int LH,
+                                int LW,
+                                int kernel_size_h,
+                                int kernel_size_w,
+                                int stride_h,
+                                int stride_w,
+                                int padding_h,
+                                int padding_w,
+                                int dilation_h,
+                                int dilation_w,
+                                tensor_view_t<4> input_tv,
+                                tensor_view_t<3> output_tv)
+{
+    /*
+     * input = {N, C, H, W}, output = {N, C * P, L}
+     * where P = kernel_size_h * kernel_size_w, L = # of blocks (see host code for
+     * formula)
+     * => gws = {ceil(N * C * P * L, LOCAL_SIZE)}, lws = {LOCAL_SIZE}
+     */
+
+    const int gid = threadIdx.x + blockIdx.x * blockDim.x;
+    int ncp = gid / L, l = gid % L;
+    int nc = ncp / P, p = ncp % P;
+    int n = nc / C, c = nc % C;
+    if(n >= N)
+        return;
+
+    int lh = l / LW, lw = l % LW;                       // sliding window position
+    int ph = p / kernel_size_w, pw = p % kernel_size_w; // position inside kernel
+    int h = lh * stride_h - padding_h + ph * dilation_h;
+    int w = lw * stride_w - padding_w + pw * dilation_w;
+
+    DTYPE x = 0;
+    if(0 <= h && h < H && 0 <= w && w < W)
+    {
+        long input_idx = input_tv.stride[3] * w + input_tv.stride[2] * h + input_tv.stride[1] * c +
+                         input_tv.stride[0] * n;
+        x = input[input_idx];
+    }
+
+    long output_idx =
+        output_tv.stride[2] * l + output_tv.stride[1] * (c * P + p) + output_tv.stride[0] * n;
```

Comment:
Same as above

Reply:
resolve in commit [f3dea16](https://github.com/moreh-dev/MIOpen/pull/37/commits/f3dea16ca398ffa36db95f75d5ef000ad2b74721)